Better Enums

joeldrapper opened the pull request:
This PR introduces a number of changes and improvements to Literal::Enums.

Literal::Enums no longer use `method_missing` to provide a "literal" syntax. Instead, you assign constants yourself.

```ruby
class Color < Literal::Enum(Integer)
  Red = new(1)
  Green = new(2)
  Blue = new(3)
end
```

The advantage to this is language servers can pick up the constants and they’ll know the types because of the call to `new`.

Every enum class has a `value` property which is specified in the class you inherit from. This will always be the first positional argument and it’s how the Enum can be efficiently coerced back and forth from a simple value for storage.

Enum classes will be frozen after definition, so you can’t add any new constants.

### Collection methods

You can look up a list of members or values:

```ruby
Color.members # => Set[Color::Red, Color::Green, Color::Blue]
```

They also enumerable so respond to `.each`, `.map`, etc.

```ruby
Color.map(&:value) # => [1, 2, 3]
```

Enumerable classes respond to `to_proc` so you can map values to enums
```ruby
[3, 2, 1].map(&Color) # => [Color::Green, Color::Blue, Color::Red]
```

### Member methods

On a given member, you can check properties, e.g.

```ruby
Color::Red.value # => 1
```

### Member singletons

You can define methods on members:

```ruby
class Switch < Literal::Enum(_Boolean)
  On = new(true) do
    def toggle = Off
  end

  Off = new(false) do
    def toggle = On
  end
end

Switch::Off.toggle # => Switch::On
```

### Custom properties

You can add your own properties

```ruby
class Color < Literal::Enum(Integer)
  prop :hex, String

  Red = new(1, hex: "#FF0000")
  Green = new(2, hex: "#00FF00")
  Blue = new(3, hex: "#0000FF")
end
```

### Indexes

You can add custom indexes

```ruby
class Country < Literal::Enum(Integer)
  prop :code, Symbol
  index :code, Symbol, unique: true

  UnitedKingdom = new(1, code: :uk)
  UnitedStates = new(2, code: :us)
end
```

Now you can look up members using `where` and `find_by`. `find_by` only works on unique indexes, while `where` works on any index.

```ruby
Country.find_by(code: :us) # => Country::UnitedStates
```

You can pass a block to create advanced indexes, e.g.

```ruby
index :country_code_length, Integer, unique: false do |country|
  country.code.length
end
```

joeldrapper commented on lib/literal/enum.rb:
```diff
@@ -1,15 +1,16 @@
 
-TracePoint.trace(:end) do |tp|
-	it = tp.self
+class Literal::Enum < Literal::Struct
```
It might make sense to extend `Literal::DataStructure` rather than `Literal::Struct` here as we definitely want different marshalling behaviour.

joeldrapper commented:
@hangsu I’d appreciate your thoughts on this. It ties together ideas from Enum and DataEnum that we explored before. I’m also thrilled with the LSP support.

![CleanShot 2024-06-25 at 14 36 58@2x](https://github.com/joeldrapper/literal/assets/246692/059bbbf0-29f5-42a6-bdef-7d45f9792a00)

hangsu commented:
I'm a big fan of everything here.

`prop` was something we wanted and actually tried to implement ourselves recently. We ultimately gave up when it got painful without forking, given the `method_missing` and `freeze`ing.

If you were pressed to reduce scope, I might suggest that `index` would be easy to replicate in business logic.

If you wanted to increase scope, I do wish `Enum` would still implement `handle`, esp since `Enum`s are difficult to use with `case/in`.

I'm very grateful that `literal` is getting some love and focus.

cc @bendillinger

joeldrapper commented:
Thanks for the feedback @hangsu! I think indexing is important for any large enumeration with multiple lookup keys, which is very common for objects like these so I’d quite like to keep it in here.

Adding back support for `handle` does depend on `Literal::Case`, which I’m not 100% convinced we want to bring back. It does have the advantage of being more exhaustive than _case-in_, so might be worth it. Also, it would be possible to keep the internals of Enum#handle private so the case objects would only be used for this.

You should be able to use Literal::Enums in case statements in a few different ways. `in Color` and `in Color::Red` were supported (though I just added new tests for them). I’ve also added support for `in Color[1]` via `deconstruct` pattern matching as well as `in Color[property: "value"]` via `deconstruct_keys`.

hangsu commented:
> in Color::Red were supported (though I just added new tests for them).

Oh, I didn't realize that, I falsely assumed only `in Color` was supported. `deconstruct` is neat. The tests help, thanks!

In that case, I would advocate for not re-implementing `Literal::Case`

eregon commented on lib/literal/enum.rb:
```diff
@@ -104,7 +149,24 @@ def name
 	end
 end
+
+TracePoint.trace(:end) do |tp|
```
FWIW, it seems very inefficient to use a TracePoint for this, because it gets called for every `class/module ... end`.
The `inherited` hook seems much better for this as then it's only called for subclasses of Literal::Enum.

eregon replied:
Ah but if it's used for freezing the class then `inherited` will be called too early I suppose.
Maybe the freezing could be done via Class#subclasses instead or so then, or an explicit call at the end of defining a Literal::Enum.

joeldrapper replied:
Unfortunately this does need to happen after the class is fully defined. The `inherited` hook is called too early.

I’m not sure how `Class#subclasses` would work here. What would be the trigger?

I’ve benchmarked this in CRuby and it doesn't seem to make much difference since `end` is only triggered a for each class and most apps don’t define that many classes. It’s usually a few thousand, not hundreds of millions and it all happens at boot.

One potential optimisation would be to create the tracepoint on `inherited` and then disable it after it is hit with the subclass it’s looking for. I believe a TracePoint can disable itself.

eregon replied:
> I’m not sure how `Class#subclasses` would work here. What would be the trigger?

Something like `Process.warmup` maybe, but it's not ideal.

Maybe this initialization could be done on the first usage of that enum lazily?
Then you wouldn't need any hook.

> I’ve benchmarked this in CRuby and it doesn't seem to make much difference since `end` is only triggered once for each class defined, and most apps don’t define that many classes. It’s usually a few thousand, not hundreds of millions and they all happen at boot.

It used to be the case that every TracePoint would disable the JIT or throw away JITed code.
`:class` and `:end` are only called from "slow/class definition" paths so they don't really need to do that, but then they also need to be handled specially and not like other TracePoints internally.
I would think creating a Binding per class and calling the TracePoint is a non-trivial overhead when an application has thousands of classes.

> One potential optimisation would be to create the tracepoint on `inherited` and then disable it after it is hit with the subclass it’s looking for. I believe a TracePoint can disable itself.

Depending on the implementation of TracePoints, enabling/disabling can be very expensive too and cause deoptimization.

---

I was also thinking maybe it could be something like:
```ruby
Color = Literal::Enum(Integer) do
  Red = new(1)
  Green = new(2)
  Blue = new(3)
end
```
but that would assign the constants to the top-level scope, so it'd need to be:
```ruby
Color = Literal::Enum(Integer) do
  self::Red = new(1)
  self::Green = new(2)
  self::Blue = new(3)
end
```
or
```ruby
Color = Literal::Enum(Integer) do
  value :Red, 1
  value :Green, 2
  value :Blue, 3
end
```
That would then naturally give you a hook after the Enum is done being defined.
(and it would also not call this hook twice if the enum class is opened in multiple places)

joeldrapper replied:
The more dynamic we get here, the less we are able to support language servers. It’s important that language servers pick up these constants and know what type they are.

fractaledmind replied:
@eregon: If you are worried about performance in a certain kind of scenario, I would recommend writing a reproducible benchmark script that demonstrates a performance issue. This will (1) prove a bottleneck exists and (2) provide a check that a "fix" indeed fixes the issue. 

Otherwise, we are talking about theoretical concerns with theoretical possible improvements.

joeldrapper replied:
We could do without the class freezing, but we would still need a trigger to generate the indexes. We never know when the final mutation occurs so we would have to generate indexes on every mutation. If you have an enum with a few hundred members, that means generating indexes a few hundred times, which is probably more overhead than waiting for the end trace point.

Also, unless I've missed something, YJIT doesn't optimise boot time. It's not even recommended to enable YJIT until after boot.

joeldrapper replied:
I benchmarked 1,000 class definitions with and without the tracepoint in CRuby. In the case without the tracepoint, I manually called `__after_defined__` at the end of the class. There’s ~34µs overhead per enum class defined.

With TracePoint: 145ms
Without TracePoint: 111ms
